README.md
<div align="center">
  <img src="./image.png" alt="banner"/>
</div>
<br>
<div align="center">
  <strong>A customizable animated checkbox component for React Native</strong>
</div>
<br>

<h2>Table of Contents</h2>
<details>
  <summary>Table of Contents</summary>
  <li><a href="#about">About</a></li>
  <li><a href="#install">Install</a></li>
  <li><a href="#usage">Usage</a></li>
  <li><a href="#props">Props</a></li>
  <li><a href="#contribute">Contribute</a></li>
  <li><a href="#license">License</a></li>
</details>

## About

A customizable animated checkbox component for React Native. No extra library required.

## Install

```sh
$ npm install --save react-native-checkbox-animated
```

```sh
$ yarn add react-native-checkbox-animated
```

## Usage

<div align="center">
  <img src="./checkbox.gif" alt="demo" />
</div>

```js
// ... Imports
import React, { useState } from "react";
import Checkbox from "react-native-checkbox-animated";

const App = () => {
  const [checked, setChecked] = useState(false);

  return (
    <View>
      ...
      <CheckBox
        label="your label here"
        onValueChange={val => setChecked(val)}
        checked={checked}
      />
      ...
    </View>
  );
};
```

Checkout examples.js for the above examples.

## Props

<table style="width:80%">
  <tr>
    <th>Property</th>
    <th>Description</th>
    <th>Required</th>
    <th>Type</th>
    <th>Default Value</th>
  </tr>
  <tr>
    <td><code>checked</code></td>
    <td>Current state of the checkBox</td>
    <td>true</td>
    <td>boolean</td>
    <td>false</td>
  </tr>
  <tr>
    <td><code>onValueChange</code></td>
    <td>Callback when checkbox is pressed. Return checkbox state</td>
    <td>true</td>
    <td>function => boolean</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>size</code></td>
    <td>Specify the size of the checkBox</td>
    <td>false</td>
    <td>number</td>
    <td>20</td>
  </tr>
  <tr>
    <td><code>label</code></td>
    <td>Text or React Component that follows the checkbox</td>
    <td>false</td>
    <td>string / React Component</td>
    <td>Your label here</td>
  </tr>
  <tr>
    <td><code>touchableLabel</code></td>
    <td>Specify if state of checkbox changes when label is pressed</td>
    <td>false</td>
    <td>boolean</td>
    <td>true</td>
  </tr>
  <tr>
    <td><code>checkedBackgroundColor</code></td>
    <td>Specify the color of checkbox's background when it's checked</td>
    <td>false</td>
    <td>string</td>
    <td>#22cdf0</td>
  </tr>
  <tr>
    <td><code>unCheckedBackgroundColor</code></td>
    <td>Specify the color of checkbox's background when it's not checked</td>
    <td>false</td>
    <td>string</td>
    <td>white</td>
  </tr>
  <tr>
    <td><code>checkedBorderColor</code></td>
    <td>Specify the color of checkbox's border when it's checked</td>
    <td>false</td>
    <td>string</td>
    <td>grey</td>
  </tr>
  <tr>
    <td><code>unCheckedBorderColor</code></td>
    <td>Specify the color of checkbox's border when it's not checked</td>
    <td>false</td>
    <td>string</td>
    <td>transparent</td>
  </tr>
  <tr>
    <td><code>borderWidth</code></td>
    <td>Border width of checkbox</td>
    <td>false</td>
    <td>number</td>
    <td>1</td>
  </tr>
  <tr>
    <td><code>rippleEffect</code></td>
    <td>Specify if label has ripple effect on press when touchableLabel is set to true</td>
    <td>false</td>
    <td>boolean</td>
    <td>true</td>
  </tr>
  <tr>
    <td><code>rippleColor</code></td>
    <td>Specify the color of ripple effect when it is set to true</td>
    <td>false</td>
    <td>string</td>
    <td>black</td>
  </tr>
  <tr>
    <td><code>checkPosition</code></td>
    <td>Specify the position where the checkbox will be rendered</td>
    <td>false</td>
    <td>enum('left' | 'right')</td>
    <td>left</td>
  </tr>
  <tr>
    <td><code>checkBoxRadius</code></td>
    <td>Specify the radius of the checkbox</td>
    <td>false</td>
    <td>number</td>
    <td>20% of size</td>
  </tr>
  <tr>
    <td><code>rounded</code></td>
    <td>Set the shape of the checkbox to be rounded instead of a square</td>
    <td>false</td>
    <td>boolean</td>
    <td>false</td>
  </tr>
  <tr>
    <td><code>checkMarkSize</code></td>
    <td>Specify the size of the checkbox</td>
    <td>false</td>
    <td>number</td>
    <td>15</td>
  </tr>
  <tr>
    <td><code>checkMarkColor</code></td>
    <td>Specify the color of the default checkmark</td>
    <td>false</td>
    <td>string</td>
    <td>black</td>
  </tr>
  <tr>
    <td><code>animationType</code></td>
    <td>Specify the custom animation type</td>
    <td>false</td>
    <td>enum('scale' | 'left' | 'reveal')</td>
    <td>scale</td>
  </tr>
  <tr>
    <td><code>checkStyle</code></td>
    <td>Text styles props applied to the checkmark character</td>
    <td>false</td>
    <td>TextStyle</td>
    <td>{}</td>
  </tr>
  <tr>
    <td><code>labelStyle</code></td>
    <td>Text styles props applied to your label</td>
    <td>false</td>
    <td>TextStyle</td>
    <td>{}</td>
  </tr>
  <tr>
    <td><code>labelContainerStyle</code></td>
    <td>View styles props applied to label container</td>
    <td>false</td>
    <td>ViewStyle</td>
    <td>{}</td>
  </tr>
  <tr>
    <td><code>containerStyle</code></td>
    <td>View styles props applied to the main container</td>
    <td>false</td>
    <td>ViewStyle</td>
    <td>{}</td>
  </tr>
  <tr>
    <td><code>checkboxContainerStyle</code></td>
    <td>View styles props applied to the container that wraps the checkbox</td>
    <td>false</td>
    <td>ViewStyle</td>
    <td>{ padding: 10 }</td>
  </tr>
  <tr>
    <td><code>boxStyle</code></td>
    <td>View styles props applied to the box that wraps the checkmark.</td>
    <td>false</td>
    <td>ViewStyle</td>
    <td>{}</td>
  </tr>
  <tr>
    <td><code>customMarker</code></td>
    <td>Replace the checkmark for your custom checkmark</td>
    <td>false</td>
    <td>ReactNode</td>
    <td>-</td>
  </tr>
</table>

## Contribute

Contributions are welcome. Please open up an issue or create PR if you would like to help out.

Note: If editing the README, please conform to the [standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## License

Licensed under the MIT License.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.10](https://github.com/MaiconGilton/react-native-checkbox-animated/compare/v0.0.9...v0.0.10) (2021-06-11)


### Bug Fixes

* remove group property ([6a15964](https://github.com/MaiconGilton/react-native-checkbox-animated/commit/6a15964a8b99f49cfd97f38ae6e3659ed9375002))

### [0.0.9](https://github.com/MaiconGilton/react-native-checkbox-animated/compare/v0.0.8...v0.0.9) (2021-06-11)


### Bug Fixes

* optional props types ([2ef4173](https://github.com/MaiconGilton/react-native-checkbox-animated/commit/2ef4173ff233d353bf6e1b4d0eaa23f3d9681137))

### 0.0.8 (2021-06-03)


### Bug Fixes

* set optional params for type checking ([00c6ae6](https://github.com/MaiconGilton/react-native-checkbox-animated/commit/00c6ae6e4dbf09eb152fb4829f10a52b76c4d330))


index.d.ts
declare module "react-native-checkbox-animated" {
    import React, { Component } from "react";
    import { StyleProp, TextStyle, ViewStyle } from "react-native";

    export interface CheckBoxProps {
        /**
         *  Current state of the checkBox. Default = false.
         */
        checked?: boolean;

        /**
         *  Callback that gets called when checkbox is pressed.
         */
        onValueChange: (val: boolean) => boolean;

        /**
         *  Specify the size of the checkBox. Default = 20
         */
        size?: number;

        /**
         *  Text that follows the checkbox.
         */
        label: string | React.ReactNode;

        /**
         *  Specify if state of checkbox changes when label is pressed. Default = true
         */
        touchableLabel?: boolean;

        /**
         *  Specify if label has ripple effect on press when touchableLabel is set to true. Default = true
         */
        rippleEffect?: boolean;

        /**
         *  Specify the color of ripple effect when it is set to true. Default = black
         */
        rippleColor?: string;

        /**
         *  Specify the color of checkbox's background when it's checked. Default = #22cdf0
         */
        checkedBackgroundColor?: string;

        /**
         *  Specify the color of checkbox's background when it's not checked. Default = white
         */
        unCheckedBackgroundColor?: string;

        /**
         *  Specify the color of checkbox's border when it's checked. Default = grey
         */
        checkedBorderColor?: string;

        /**
         *  Specify the color of checkbox's border when it's not checked. Default = transparent
         */
        unCheckedBorderColor?: string;

        /**
         *  border width of checkbox. Default = 1
         */
        borderWidth?: boolean;

        /**
         *  Specify the position where the checkbox will be rendered. Default = left
         */
        checkPosition?: "left" | "right";

        /**
         *  Set the shape of the checkbox to be rounded instead of a square. Default = false
         */
        rounded?: boolean;

        /**
         *  Set radius of the checkbox. Default = 20% of size
         */
        checkBoxRadius?: number;

        /**
         *  Specify the size of the checkbox. Default = 15
         */
        checkMarkSize?: number;

        /**
         *  Specify the color of the checkmark. Default = black
         *  Important: checkmark 1 has a fixed grey color, you cannot change it.
         */
        checkMarkColor?: string;

        /**
         *  Specify the custom animation type. Default = scale
         */
        animationType?: "scale" | "left" | "reveal";

        /**
         *  Text styles props applied to the checkmark character.
         */
        checkStyle?: StyleProp<TextStyle>;

        /**
         *  Text styles props applied to your label.
         */
        labelStyle?: StyleProp<TextStyle>;

        /**
         *  View styles props applied to label container.
         */
        labelContainerStyle?: StyleProp<ViewStyle>;

        /**
         *  View styles props applied to the main container.
         */
        containerStyle?: StyleProp<ViewStyle>;

        /**
         *  View styles props applied to the container that wraps the checkbox.
         */
        checkboxContainerStyle?: StyleProp<ViewStyle>;

        /**
         *  View styles props applied to the box that wraps the checkmark.
         */
        boxStyle?: StyleProp<ViewStyle>;

        /**
         *  Replace the checkmark for your custom mark.
         */
        customMarker?: () => React.ReactNode;
    }

    class CheckBox extends Component<CheckBoxProps> { }

    export default CheckBox;
}


CheckBox.js
import React from "react";
import {
  TouchableOpacity,
  Text,
  View,
  Animated,
  Easing,
  Platform
} from "react-native";

export default class CheckBox extends React.Component {
  animationScale = new Animated.Value(this.props.checked ? 1 : 0);
  animationLeft = new Animated.Value(
    this.props.checked ? 0 : -this.props.size
  );
  animationReveal = new Animated.Value(
    this.props.checked ? this.props.size : 0
  );
  rippleScale = new Animated.Value(0.01);
  rippleOpacity = new Animated.Value(0.1);

  componentDidUpdate() {
    const { checked, animationType, group } = this.props;

    if (group && !checked) {
      if (animationType === "scale") this.animateScale(true);
      else if (animationType === "left") this.animateLeft(true);
      else this.animateReveal(true);
    }
  }

  _onValueChange = () => {
    const { checked, onValueChange, animationType, group } = this.props;

    if (group && checked) return;

    onValueChange(!checked);

    if (animationType === "scale") this.animateScale(checked);
    else if (animationType === "left") this.animateLeft(checked);
    else this.animateReveal(checked);
  };

  animateScale = (checked) => {
    if (checked)
      Animated.timing(this.animationScale, {
        toValue: 0.01,
        duration: 100,
        easing: Easing.bezier(0.0, 0.0, 0.2, 1),
        useNativeDriver: true
      }).start();
    else {
      Animated.timing(this.animationScale, {
        toValue: 1,
        easing: Easing.elastic(2),
        duration: 100,
        useNativeDriver: true
      }).start();
    }
  };

  animateLeft = (checked) => {
    const { size } = this.props;

    if (checked)
      Animated.timing(this.animationLeft, {
        toValue: -size,
        duration: 50,
        easing: Easing.ease,
        useNativeDriver: true
      }).start();
    else
      Animated.timing(this.animationLeft, {
        toValue: 0,
        easing: Easing.elastic(1.2),
        duration: 150,
        useNativeDriver: true
      }).start();
  };

  animateReveal = (checked) => {
    const { size } = this.props;

    if (checked)
      Animated.timing(this.animationReveal, {
        toValue: 0,
        duration: 50,
        easing: Easing.ease,
        useNativeDriver: true
      }).start();
    else
      Animated.timing(this.animationReveal, {
        toValue: size,
        easing: Easing.ease,
        duration: 100,
        useNativeDriver: true
      }).start();
  };

  _renderTextBtn = (position) => {
    const {
      label,
      touchableLabel,
      checkPosition,
      labelStyle,
      rippleEffect,
      labelContainerStyle
    } = this.props;

    if (!label || position !== checkPosition) return null;

    return (
      <TouchableOpacity
        onPressIn={rippleEffect && touchableLabel ? this.onPressedIn : () => { }}
        onPressOut={
          rippleEffect && touchableLabel ? this.onPressedOut : () => { }
        }
        onPress={touchableLabel ? this._onValueChange : () => { }}
        activeOpacity={touchableLabel ? 0.7 : 1}
        style={[{ justifyContent: "center", flex: 1 }, labelContainerStyle]}
      >
        <Text style={[{ padding: 10 }, labelStyle]}>{label}</Text>
      </TouchableOpacity>
    );
  };

  _renderCheckBtn = () => {
    const {
      checked,
      size,
      checkedBackgroundColor,
      unCheckedBackgroundColor,
      unCheckedBorderColor,
      checkedBorderColor,
      borderWidth,
      rounded,
      checkBoxRadius,
      checkMarkSize,
      checkMarkColor,
      animationType,
      checkStyle,
      checkboxContainerStyle,
      boxStyle,
      customMarker,
    } = this.props;

    var animate = {};
    if (animationType === "scale")
      animate = { transform: [{ scale: this.animationScale }] };
    else if (animationType === "left")
      animate = { transform: [{ translateX: this.animationLeft }] };

    return (
      <TouchableOpacity
        onPress={this._onValueChange}
        activeOpacity={0.7}
        style={[{ padding: 10 }, checkboxContainerStyle]}
      >
        <Animated.View
          style={[
            {
              width: size,
              height: size,
              borderRadius: rounded ? size : checkBoxRadius || size * 0.2,
              borderWidth: borderWidth,
              borderColor: checked ? checkedBorderColor : unCheckedBorderColor,
              justifyContent: "center",
              alignItems: "center",
              overflow: "hidden",
              backgroundColor: checked
                ? checkedBackgroundColor
                : unCheckedBackgroundColor
            },
            boxStyle
          ]}
        >
          {customMarker ? (
            <Animated.View style={animate}>{customMarker}</Animated.View>
          ) : (
              <Animated.Text
                style={[
                  {
                    fontSize: checkMarkSize,
                    lineHeight: size,
                    color: checkMarkColor
                  },
                  animate,
                  checkStyle
                ]}
              >✓</Animated.Text>
            )}

          {animationType === "reveal" ? (
            <Animated.View
              style={{
                position: "absolute",
                width: size,
                aspectRatio: 1 / 1,
                borderRadius: rounded ? size : size * 0.05,
                backgroundColor: checked
                  ? checkedBackgroundColor
                  : unCheckedBackgroundColor,
                transform: [{ translateX: this.animationReveal }]
              }}
            />
          ) : null}
        </Animated.View>
      </TouchableOpacity>
    );
  };

  onPressedIn = () => {
    Animated.timing(this.rippleScale, {
      toValue: 1,
      duration: 250,
      easing: Easing.bezier(0.0, 0.0, 0.05, 1),
      useNativeDriver: Platform.OS === "android"
    }).start();
  };

  onPressedOut = () => {
    Animated.timing(this.rippleOpacity, {
      toValue: 0,
      duration: 50,
      useNativeDriver: Platform.OS === "android"
    }).start(() => {
      this.rippleScale.setValue(0.01);
      this.rippleOpacity.setValue(0.1);
    });
  };

  _renderRipple() {
    return (
      <Animated.View
        style={{
          position: "absolute",
          top: 0,
          left: 0,
          bottom: 0,
          right: 0,
          transform: [{ scale: this.rippleScale }],
          opacity: this.rippleOpacity,
          backgroundColor: this.props.rippleColor
        }}
      />
    );
  }

  render() {
    return (
      <View
        style={[
          { flexDirection: "row", alignItems: "center" },
          this.props.containerStyle
        ]}
      >
        {this._renderRipple()}
        {this._renderTextBtn("right")}
        {this._renderCheckBtn()}
        {this._renderTextBtn("left")}
      </View>
    );
  }
}

CheckBox.defaultProps = {
  label: 'Your label here',
  checked: false,
  touchableLabel: true,
  size: 20,
  checkPosition: 'left',
  checkedBackgroundColor: "#22cdf0",
  unCheckedBackgroundColor: "white",
  unCheckedBorderColor: "grey",
  checkedBorderColor: "transparent",
  borderWidth: 1,
  rippleEffect: true,
  rippleColor: 'black',
  rounded: false,
  checkMarkSize: 12,
  checkMarkColor: "black",
  animationType: "scale",
}

examples.js
import React, { useState } from 'react';
import {
    SafeAreaView,
    ScrollView,
    Text,
    StatusBar,
    View,
} from 'react-native';
import CheckBox from 'react-native-checkbox-animated'

const App = () => {

    const [check1, setcheck1] = useState(false)
    const [check2, setcheck2] = useState(false)
    const [check3, setcheck3] = useState(false)
    const [check4, setcheck4] = useState(false)
    const [option, setoption] = useState(1)

    return (
        <SafeAreaView style={{ flex: 1, }}>
            <StatusBar barStyle="dark-content" />
            <ScrollView contentContainerStyle={{ margin: 20 }}>
                <Text style={{ fontSize: 18, marginVertical: 20, textAlign: 'center' }}>React Native Animated CheckBox</Text>

                <CheckBox
                    label='CheckBox position left and no ripple effect'
                    onValueChange={(val) => setcheck1(val)}
                    checked={check1}
                    // checkMarkSize={20}
                    animationType='left'
                    checkMarkColor='white'
                    // boxContainerStyle={{ borderRadius: 10 }}
                    rippleEffect={false}
                />

                <CheckBox
                    label='CheckBox position right and custom size'
                    onValueChange={(val) => setcheck2(val)}
                    checked={check2}
                    checkPosition='right'
                    size={30}
                    checkMarkSize={25}
                />

                <CheckBox
                    label='CheckBox round and custom style'
                    checkPosition='right'
                    onValueChange={(val) => setcheck3(val)}
                    checked={check3}
                    size={25}
                    checkMarkSize={15}
                    checkMarkColor={'white'}
                    checkedBackgroundColor={'green'}
                    // checkedBackgroundColor={'white'}
                    rounded
                />

                <CheckBox
                    label='Custom marker'
                    onValueChange={(val) => setcheck4(val)}
                    checked={check4}
                    size={30}
                    checkedBackgroundColor='white'
                    unCheckedBackgroundColor='transparent'
                    customMarker={<Text style={{ fontSize: 22 }}>☺</Text>}
                />

                <View style={{ borderRadius: 10, borderWidth: 1, padding: 10, marginVertical: 20 }}>
                    <Text style={{ fontSize: 16 }}>Grouped checkboxes</Text>

                    <CheckBox
                        label='option 1'
                        onValueChange={(val) => setoption(1)}
                        checked={option === 1}
                        containerStyle={{}}
                        animationType='left'
                    />

                    <CheckBox
                        label='option 2'
                        onValueChange={(val) => setoption(2)}
                        animationType='left'
                        checked={option === 2}
                    />

                    <CheckBox
                        label='option 3'
                        onValueChange={(val) => setoption(3)}
                        animationType='left'
                        checked={option === 3}
                    />
                </View>
            </ScrollView>
        </SafeAreaView>
    );
};

export default App;


index.js
import React from "react";
import {
    TouchableOpacity,
    Text,
    View,
    Animated,
    Easing,
    Platform
} from "react-native";

export default class CheckBox extends React.Component {
    state = { checked: this.props.checked }
    animationScale = new Animated.Value(this.props.checked ? 1 : 0);
    animationLeft = new Animated.Value(
        this.state.checked ? 0 : -this.props.size
    );
    animationReveal = new Animated.Value(
        this.state.checked ? this.props.size : 0
    );
    rippleScale = new Animated.Value(0.01);
    rippleOpacity = new Animated.Value(0.1);

    componentDidUpdate(prevProps, prevState) {
        const { animationType, checked } = this.props;

        if (animationType === "scale") this.animateScale(!checked);
        else if (animationType === "left") this.animateLeft(!checked);
        else this.animateReveal(!checked);

        if (checked !== prevProps.checked) this.setState({ checked })
    }

    _onValueChange = () => {
        const { onValueChange, animationType } = this.props;

        onValueChange(!this.state.checked);

        if (animationType === "scale") this.animateScale(this.state.checked);
        else if (animationType === "left") this.animateLeft(this.state.checked);
        else this.animateReveal(this.state.checked);
    };

    animateScale = (checked) => {
        if (checked)
            Animated.timing(this.animationScale, {
                toValue: 0.01,
                duration: 100,
                easing: Easing.bezier(0.0, 0.0, 0.2, 1),
                useNativeDriver: true
            }).start();
        else {
            Animated.timing(this.animationScale, {
                toValue: 1,
                easing: Easing.elastic(2),
                duration: 100,
                useNativeDriver: true
            }).start();
        }
    };

    animateLeft = (checked) => {
        const { size } = this.props;

        if (checked)
            Animated.timing(this.animationLeft, {
                toValue: -size,
                duration: 50,
                easing: Easing.ease,
                useNativeDriver: true
            }).start();
        else
            Animated.timing(this.animationLeft, {
                toValue: 0,
                easing: Easing.elastic(1.2),
                duration: 150,
                useNativeDriver: true
            }).start();
    };

    animateReveal = (checked) => {
        const { size } = this.props;

        if (checked)
            Animated.timing(this.animationReveal, {
                toValue: 0,
                duration: 50,
                easing: Easing.ease,
                useNativeDriver: true
            }).start();
        else
            Animated.timing(this.animationReveal, {
                toValue: size,
                easing: Easing.ease,
                duration: 100,
                useNativeDriver: true
            }).start();
    };

    _renderTextBtn = (position) => {
        const {
            label,
            touchableLabel,
            checkPosition,
            labelStyle,
            rippleEffect,
            labelContainerStyle
        } = this.props;

        if (!label || position !== checkPosition) return null;

        return (
            <TouchableOpacity
                onPressIn={rippleEffect && touchableLabel ? this.onPressedIn : () => { }}
                onPress={() => {
                    if (touchableLabel) {
                        this.setState({ checked: !this.state.checked })
                        this._onValueChange(!this.state.checked)
                    }
                }}
                activeOpacity={touchableLabel ? 0.7 : 1}
                style={[{ justifyContent: "center", flex: 1 }, labelContainerStyle]}
            >
                {React.isValidElement(label) ? label : <Text style={[{ padding: 10 }, labelStyle]}>{label}</Text>}
            </TouchableOpacity>
        );
    };

    _renderCheckBtn = () => {
        const {
            size,
            checkedBackgroundColor,
            unCheckedBackgroundColor,
            unCheckedBorderColor,
            checkedBorderColor,
            borderWidth,
            rounded,
            checkBoxRadius,
            checkMarkSize,
            checkMarkColor,
            animationType,
            checkStyle,
            checkboxContainerStyle,
            boxStyle,
            customMarker,
        } = this.props;

        const { checked } = this.state

        var animate = {};
        if (animationType === "scale")
            animate = { transform: [{ scale: this.animationScale }] };
        else if (animationType === "left")
            animate = { transform: [{ translateX: this.animationLeft }] };

        return (
            <TouchableOpacity
                onPress={() => {
                    this.setState({ checked: !this.state.checked })
                    this._onValueChange(!this.state.checked)
                }}
                activeOpacity={0.7}
                style={[{ padding: 10 }, checkboxContainerStyle]}
            >
                <Animated.View
                    style={[
                        {
                            width: size,
                            height: size,
                            borderRadius: rounded ? size : checkBoxRadius || size * 0.2,
                            borderWidth: borderWidth,
                            borderColor: checked ? checkedBorderColor : unCheckedBorderColor,
                            justifyContent: "center",
                            alignItems: "center",
                            overflow: "hidden",
                            backgroundColor: checked
                                ? checkedBackgroundColor
                                : unCheckedBackgroundColor
                        },
                        boxStyle
                    ]}
                >
                    {customMarker ? (
                        <Animated.View style={animate}>{customMarker}</Animated.View>
                    ) : (
                        <Animated.Text
                            style={[
                                {
                                    fontSize: checkMarkSize,
                                    lineHeight: size,
                                    color: checkMarkColor
                                },
                                animate,
                                checkStyle
                            ]}
                        >✓</Animated.Text>
                    )}

                    {animationType === "reveal" ? (
                        <Animated.View
                            style={{
                                position: "absolute",
                                width: size,
                                aspectRatio: 1 / 1,
                                borderRadius: rounded ? size : size * 0.05,
                                backgroundColor: checked
                                    ? checkedBackgroundColor
                                    : unCheckedBackgroundColor,
                                transform: [{ translateX: this.animationReveal }]
                            }}
                        />
                    ) : null}
                </Animated.View>
            </TouchableOpacity>
        );
    };

    onPressedIn = () => {
        Animated.parallel([
            Animated.timing(this.rippleScale, {
                toValue: 1,
                duration: 150,
                easing: Easing.inOut(Easing.quad),
                useNativeDriver: Platform.OS === "android"
            }),
            Animated.timing(this.rippleOpacity, {
                toValue: 0,
                duration: 300,
                easing: Easing.inOut(Easing.quad),
                useNativeDriver: Platform.OS === "android"
            })
        ]).start(() => {
            this.rippleScale.setValue(0.01);
            this.rippleOpacity.setValue(0.1);
        });
    };

    _renderRipple() {
        return (
            <Animated.View
                style={{
                    position: "absolute",
                    top: 0,
                    left: 0,
                    bottom: 0,
                    right: 0,
                    transform: [{ scale: this.rippleScale }],
                    opacity: this.rippleOpacity,
                    backgroundColor: this.props.rippleColor
                }}
            />
        );
    }

    render() {
        return (
            <View
                style={[
                    { flexDirection: "row", alignItems: "center" },
                    this.props.containerStyle
                ]}
            >
                {this._renderRipple()}
                {this._renderTextBtn("right")}
                {this._renderCheckBtn()}
                {this._renderTextBtn("left")}
            </View>
        );
    }
}

CheckBox.defaultProps = {
    label: 'Your label here',
    checked: false,
    touchableLabel: true,
    size: 20,
    checkPosition: 'left',
    checkedBackgroundColor: "#22cdf0",
    unCheckedBackgroundColor: "white",
    unCheckedBorderColor: "grey",
    checkedBorderColor: "transparent",
    borderWidth: 1,
    rippleEffect: true,
    rippleColor: 'black',
    rounded: false,
    checkMarkSize: 12,
    checkMarkColor: "black",
    animationType: "scale",
}